lib/IRGen/IRGenSIL.cpp

```diff
@@ -2024,43 +2012,79 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+  // TODO: We should also handle the following cases:
+  // - if `opInput` is an array of TensorFlow values, then we add all the
+  //   elements' handles;
+  // - if `opInput` is an array of archetypes conforming to TensorGroup,
+  //   then we ask the conformance for all the elements' handles, flatten the
+  //   results together, and add those;
+  auto unpackAndAddInput = [&](SILValue opInput) -> llvm::Value* {
     LLVM_DEBUG(llvm::dbgs()
-               << " Adding input of type " << tensorHandle->getType() << ".\n");
-    assert(tf::isTensorFlowValue(tensorHandle->getType()) &&
-           "all op inputs should be TensorFlow values");
-
-    auto tensorHandleValue = getLoweredSingletonExplosion(tensorHandle);
-    auto *extractHandleFn = IGM.getTFC_GetCTensorHandleFromSwiftFn();
-    // Cast to a void*, as required by TFC_GetCTensorHandleFromSwift().
-    auto tensorHandleValueUntyped =
+               << " Adding input of type " << opInput->getType() << ".\n");
+
+    if (tf::isTensorFlowValue(opInput->getType())) {
+      auto *tensorHandleValue = getLoweredSingletonExplosion(opInput);
+      auto *opAddInputFromTensorHandleFn =
+          IGM.getTFC_OpAddInputFromTensorHandleFn();
+      auto *tensorHandleValueUntyped =
+          Builder.CreateBitCast(tensorHandleValue, IGM.Int8PtrTy);
+      Builder.CreateCall(opAddInputFromTensorHandleFn,
+                         {op, tensorHandleValueUntyped, status});
+      checkOk(status);
+      return llvm::ConstantInt::get(IGM.Int32Ty, 1);
+    }
+
+    auto canType = opInput->getType().getASTType()->getCanonicalType();
+    auto conformance = tfModule->lookupConformance(canType,
+                                                   tensorGroupProto);
+    assert(conformance && "input type does not conform to TensorGroup");
+    auto *typeMetadata = emitTypeMetadataRef(canType);
+    auto *wtable = emitWitnessTableRef(*this, canType, *conformance);
+
+    auto *tensorHandleValue = getLoweredAddress(opInput).getAddress();
+    auto *tensorHandleValueUntyped =
         Builder.CreateBitCast(tensorHandleValue, IGM.Int8PtrTy);
-    auto cHandle =
-        Builder.CreateCall(extractHandleFn, {tensorHandleValueUntyped});
 
-    // Add an op input as in:
-    //   TFE_OpAddInput(op, cHandle);
-    auto *opAddInputFn = IGM.getTFE_OpAddInputFn();
-    Builder.CreateCall(opAddInputFn, {op, cHandle, status});
+    auto *opAddInputFromTensorGroupFn =
+        IGM.getTFC_OpAddInputFromTensorGroupFn();
+    auto *ret = Builder.CreateCall(
+        opAddInputFromTensorGroupFn,
+        {op, tensorHandleValueUntyped, status, typeMetadata, wtable});
     checkOk(status);
+    return ret;
   };
 
+  // If we find an out parameter, we'll keep track of its address and some data
+  // about its type.
+  Address outParameterAddress;
+  llvm::Value *outParameterTypeMetadata = nullptr;
+  llvm::Value *outParameterTensorGroupWitnessTable = nullptr;
+
   // Handle inputs and dynamic attributes.
+  unsigned inputIdx = 0;
```

Comment:
with the packing and unpacking going on, the semantics `inputIdx` could become unclear, and the associated code could be harder to read and maintain.

consider adding an example, and maybe introduce some additional terminology. one option can be (pls feel free to come up with sth else):

say the op takes 3 _input params_ (with additional attrs but we don't consider them input), where each param can be a single tensor handle or a list. Say this op takes: (handle, list, handle), and the list in this case contains 3 _elements_.

We first unpack these params into a flat list of 5 tensor handle _elements_. For the middle _param_, it has an attr N, and we set it to be 3.

Reply:
I added a comment with an example that should clarify what is happening.

Reply:
very nice